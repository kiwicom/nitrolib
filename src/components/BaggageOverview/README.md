# BaggageOverview

- renders baggage overview
- depending on props can be used as standalone component or wrapped in [Container ](`./components/Container`)

**Example:**

standalone

```js
<BaggageOverview
  definitions={definitions}
  FAQLinksHandler={category => {}}
  context="MMB-PassengerCard"
/>
```

wrapped in Container

```js
<Container
  passengers={[
    {
      paxId: 3,
      firstName: "George",
      lastName: "Bush",
      baggage: {
        holdBag: 0,
        handBag: 1
      }
    }
  ]}
  baggage={baggageData}
  context="booking"
>
  {({ props }) => <BaggageOverview {...props} />}
</Container>
```
