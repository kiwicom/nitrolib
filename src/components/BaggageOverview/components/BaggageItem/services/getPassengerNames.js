// @flow strict

import type { BaggagePassengerType } from "../../../../../records/Baggage";

const getPassengerNames = (passengersArr: BaggagePassengerType[]): string =>
  passengersArr
    .map(p => `${p.firstName[0]}. ${p.middleName ? `${p.middleName[0]}. ` : ""}${p.lastName}`)
    .join(", ");

export default getPassengerNames;
