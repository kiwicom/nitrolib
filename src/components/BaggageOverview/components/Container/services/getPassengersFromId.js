// @flow strict
import type { BaggagePassengerType, Passenger } from "../../../../../records/Baggage";

export default function getPassengersFromId(
  paxIds: number[],
  passengersArray: Passenger[],
): BaggagePassengerType[] {
  return paxIds
    .map(paxId => {
      const passenger: ?Passenger = passengersArray.find(p => p.paxId === paxId);
      return (
        passenger && {
          paxId: passenger.paxId,
          firstName: passenger.firstName,
          middleName: passenger.middleName,
          lastName: passenger.lastName,
        }
      );
    })
    .filter(Boolean);
}
