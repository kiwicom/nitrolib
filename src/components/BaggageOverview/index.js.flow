// @flow strict
import type {
  FAQLinksHandlerType,
  DefinitionWithPassenger,
  HoldBagDefinitionWithId,
  HandBagDefinitionWithId,
  OverviewContextType,
} from "../../records/Baggage";

type Props = {|
  definitions?: (HandBagDefinitionWithId | HoldBagDefinitionWithId)[],
  definitionsWithPassengers?: DefinitionWithPassenger[],
  FAQLinksHandler?: FAQLinksHandlerType,
  context: OverviewContextType,
|};

declare export default React.ComponentType<Props>;
