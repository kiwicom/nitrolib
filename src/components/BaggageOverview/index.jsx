// @flow
import * as React from "react";
import styled, { css } from "styled-components";
import R from "ramda";
import Text from "@kiwicom/orbit-components/lib/Text";
import Stack from "@kiwicom/orbit-components/lib/Stack";
import BaggagePersonalItemNone from "@kiwicom/orbit-components/lib/icons/BaggagePersonalItemNone";
import mq from "@kiwicom/orbit-components/lib/utils/mediaQuery";

import { themeDefault } from "../../records/Theme";
import type { ThemeProps } from "../../records/Theme";
import type {
  FAQLinksHandlerType,
  DefinitionWithPassenger,
  HoldBagDefinitionWithId,
  HandBagDefinitionWithId,
  OverviewContextType,
} from "../../records/Baggage";
import BaggageItem from "./components/BaggageItem";
import Translate from "../Translate/index";

type Props = {
  definitions?: Array<HandBagDefinitionWithId | HoldBagDefinitionWithId>,
  definitionWithPassengers?: Array<DefinitionWithPassenger>,
  FAQLinksHandler?: FAQLinksHandlerType,
  context: OverviewContextType,
};

type WrapperProps = ThemeProps & {
  context: OverviewContextType,
};
const Wrapper = styled.div`
  width: 100%;
  > * {
    margin-bottom: 10px;
    ${mq.mediumMobile(css`
      margin-bottom: ${({ theme, context }: WrapperProps) =>
        context === "MMB-PassengersSummary" ? theme.orbit.spaceXSmall : theme.orbit.spaceXXSmall};
    `)};
  }
`;
Wrapper.defaultProps = {
  theme: themeDefault,
  context: "booking",
};

const NoPersonalItem = () => (
  <Stack shrink spacing="condensed" align="center">
    <BaggagePersonalItemNone size="medium" color="primary" />
    <Text>
      <Translate t="baggage_modal.select.no_personal_item" />
    </Text>
  </Stack>
);

const BaggageOverview = ({
  definitions,
  definitionWithPassengers,
  context,
  FAQLinksHandler,
}: Props) => {
  const getDefinitions = () => {
    if (definitionWithPassengers) {
      return definitionWithPassengers;
    }
    if (definitions) {
      const groupedDefinitions = definitions.reduce((acc, def) => {
        if (acc[def.id]) {
          acc[def.id].amount += 1;
        } else {
          acc[def.id] = {
            amount: 1,
            ...def,
          };
        }
        return acc;
      }, {});
      return R.values(groupedDefinitions);
    }
    return null;
  };
  const baggages = getDefinitions();
  return (
    <Wrapper context={context}>
      {baggages && baggages.some(bag => bag.category === "personalItem") && <NoPersonalItem />}
      {baggages &&
        baggages.map((bag, index) => (
          <BaggageItem
            key={index} // eslint-disable-line
            amount={bag.passengers ? bag.passengers.length : bag.amount}
            passengers={bag.passengers}
            category={bag.category}
            restrictions={bag.restrictions}
            FAQLinksHandler={FAQLinksHandler}
            context={context}
          />
        ))}
    </Wrapper>
  );
};

export default BaggageOverview;
