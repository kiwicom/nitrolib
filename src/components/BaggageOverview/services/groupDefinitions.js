// @flow strict
import * as R from "ramda";

import type { Definition } from "../../../records/Baggage";

export default function groupDefinitions(definitions: Definition[]) {
  const groupedDefinitions = definitions.reduce((acc, def) => {
    if (acc[def.index]) {
      acc[def.index].amount += 1;
    } else {
      acc[def.index] = {
        amount: 1,
        ...def,
      };
    }
    return acc;
  }, {});
  return R.values(groupedDefinitions);
}
