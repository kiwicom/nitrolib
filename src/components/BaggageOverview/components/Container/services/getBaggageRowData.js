// @flow strict
import * as R from "ramda";

import getPassengersFromId from "./getPassengersFromId";
import type {
  BaggageCategory,
  Passenger,
  BaggageType,
  DefinitionWithPassenger,
} from "../../../../../records/Baggage";

type getBaggageRowDataArgs = {
  bagType: BaggageCategory,
  passengers: Passenger[],
  baggage: BaggageType,
};

export default function getBaggageRowData({
  bagType,
  baggage,
  passengers,
}: getBaggageRowDataArgs): DefinitionWithPassenger[] {
  const { combinations } = baggage;
  const definitions = baggage.definitions[bagType];

  const passengersWithBagDefinitionsIndices = passengers.map(passenger => ({
    paxId: passenger.paxId,
    definitionsIndices: {
      handBag: combinations.handBag[passenger.baggage.handBag].indices,
      holdBag: combinations.holdBag[passenger.baggage.holdBag].indices,
    },
  }));

  const baggageData = passengersWithBagDefinitionsIndices.reduce(
    (acc, passenger) =>
      acc.concat(
        ...passenger.definitionsIndices[bagType].map(bagIndex => ({
          paxId: passenger.paxId,
          originalIndex: bagIndex,
          category: definitions[bagIndex].category,
          restrictions: definitions[bagIndex].restrictions,
        })),
      ),
    [],
  );
  const baggageWithPassengersIds = baggageData.reduce((acc, bag) => {
    const bagPassengers = acc[bag.originalIndex]?.passengers || [];
    acc[bag.originalIndex] = {
      originalIndex: bag.originalIndex,
      category: bag.category,
      passengers: [...bagPassengers, bag.paxId],
      restrictions: bag.restrictions,
    };
    return acc;
  }, {});

  return R.values(baggageWithPassengersIds).map(bag => ({
    ...bag,
    passengers: getPassengersFromId(bag.passengers, passengers),
  }));
}
