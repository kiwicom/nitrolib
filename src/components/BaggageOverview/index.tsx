import * as React from "react";
import * as R from "ramda";
import Stack from "@kiwicom/orbit-components/lib/Stack";

import groupDefinitions from "./services/groupDefinitions";
import {
  FAQLinksHandlerType,
  DefinitionWithPassenger,
  Definition,
  OverviewContextType,
} from "../../records/Baggage";
import BaggageItem from "./components/BaggageItem";
import NoPersonalItem from "./components/NoPersonalItem";

type Props = {
  definitions?: Definition[],
    definitionsWithPassengers ?: DefinitionWithPassenger[],
    FAQLinksHandler ?: FAQLinksHandlerType,
    context: OverviewContextType,
};

const BaggageOverview = ({
  definitions,
  definitionsWithPassengers,
  context,
  FAQLinksHandler,
}: Props) => {
  const baggages = definitionsWithPassengers || (definitions && groupDefinitions(definitions));

  return (
    <Stack
      flex
      direction="column"
      dataTest={`BaggageOverview-${context}`}
      mediumMobile={{
        spacing: context === "MMB-PassengersSummary" ? "condensed" : "tight",
        direction: "column",
      }}
    >
      {baggages && !baggages.some(R.propEq("category", "personalItem")) && <NoPersonalItem />}
      {baggages &&
        baggages.map((bag, index) => (
          <BaggageItem
            key={index} // eslint-disable-line
            amount={bag.passengers ? bag.passengers.length : bag.amount}
            passengers={bag.passengers}
            category={bag.category}
            restrictions={bag.restrictions}
            FAQLinksHandler={FAQLinksHandler}
            context={context}
          />
        ))}
    </Stack>
  );
};

export default BaggageOverview;
