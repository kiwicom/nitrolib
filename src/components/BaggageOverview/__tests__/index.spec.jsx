// @flow strict
import * as React from "react";
import { shallow } from "enzyme";

import { baggageData } from "../../BaggagePicker/services/data";
import BaggageOverview from "../index";

const passengers = [
  {
    id: 1,
    firstName: "Barrack",
    lastName: "Obama",
    baggage: {
      holdBag: 1,
      handBag: 3,
    },
  },
  {
    id: 2,
    firstName: "Donald",
    lastName: "Trump",
    baggage: {
      holdBag: 1,
      handBag: 4,
    },
  },
  {
    id: 3,
    firstName: "George",
    lastName: "Bush",
    baggage: {
      holdBag: 0,
      handBag: 2,
    },
  },
];

const props = {
  passengers,
  baggage: baggageData,
  currentPassengerId: undefined,
};

describe("#BaggageOverview", () => {
  test("render ", () => {
    const wrapper = shallow(<BaggageOverview {...props} />);
    expect(wrapper.find("BaggageOverview__Wrapper").exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });
});
