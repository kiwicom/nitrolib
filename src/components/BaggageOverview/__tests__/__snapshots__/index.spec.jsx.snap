// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#BaggageOverview render  1`] = `
<BaggageOverview__Wrapper
  spacing="tight"
>
  <Text>
    Baggage
  </Text>
  <BaggageItem
    amount={2}
    category="personalItem"
    isPassengersShowed={true}
    key="0"
    passengers={
      Array [
        Object {
          "firstName": "Barrack",
          "id": 1,
          "lastName": "Obama",
        },
        Object {
          "firstName": "Donald",
          "id": 2,
          "lastName": "Trump",
        },
      ]
    }
    restrictions={
      Object {
        "dimensionsSum": null,
        "height": 20,
        "length": 20,
        "weight": 5,
        "width": 20,
      }
    }
  />
  <BaggageItem
    amount={4}
    category="cabinBag"
    isPassengersShowed={true}
    key="1"
    passengers={
      Array [
        Object {
          "firstName": "Barrack",
          "id": 1,
          "lastName": "Obama",
        },
        Object {
          "firstName": "Donald",
          "id": 2,
          "lastName": "Trump",
        },
        Object {
          "firstName": "George",
          "id": 3,
          "lastName": "Bush",
        },
      ]
    }
    restrictions={
      Object {
        "dimensionsSum": null,
        "height": 35,
        "length": 45,
        "weight": 10,
        "width": 20,
      }
    }
  />
  <BaggageItem
    amount={2}
    category="holdBag"
    isPassengersShowed={true}
    key="0"
    passengers={
      Array [
        Object {
          "firstName": "Barrack",
          "id": 1,
          "lastName": "Obama",
        },
        Object {
          "firstName": "Donald",
          "id": 2,
          "lastName": "Trump",
        },
      ]
    }
    restrictions={
      Object {
        "dimensionsSum": 156,
        "height": 52,
        "length": 78,
        "weight": 10,
        "width": 26,
      }
    }
  />
</BaggageOverview__Wrapper>
`;
